File: templates/go-aoc/internal/get-aoc-input/scaffold.go
package aocinput

import (
	"fmt"
	"os"
	"path/filepath"
)

const (
	taskFile = "main.go"
	taskGo   = `package main

import (
	"fmt"
	"time"

	"aoc/internal/parse-input"
)

func part1(lines []string) int {
	// TODO: implement
	return 0
}

func part2(lines []string) int {
	// TODO: implement
	return 0
}

func main() {
	start := time.Now()

	parseStart := time.Now()
	input, err := parse.Lines("./input.txt")
	if err != nil {
		panic(err)
	}
	fmt.Printf("parse-input: %s\n", time.Since(parseStart))

	p1Start := time.Now()
	fmt.Println("PART 1: ", part1(input))
	fmt.Printf("part 1: %s\n", time.Since(p1Start))

	p2Start := time.Now()
	fmt.Println("PART 2: ", part2(input))
	fmt.Printf("part 2: %s\n", time.Since(p2Start))

	fmt.Printf("task: %s\n", time.Since(start))
}
`
	readmeGo = `[%[2]s](https://adventofcode.com/%[1]d/day/%[3]d "%[2]s")

` + "```shell\n" + `go run ./tasks/day%[3]d
` + "```\n"
)

type WriteTaskInput struct {
	OutputDir string
	Data      string
}

func WriteTask(in WriteTaskInput) error {
	return os.WriteFile(filepath.Join(in.OutputDir, "input.txt"), []byte(in.Data), 0o644)
}

type WriteTemplateInput struct {
	OutputDir string
	Year      int
	Day       int
	Title     string
}

func WriteTemplate(in WriteTemplateInput) error {
	if err := os.WriteFile(filepath.Join(in.OutputDir, "README.md"), fmt.Appendf(nil, readmeGo, in.Year, in.Title, in.Day), 0o644); err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(in.OutputDir, taskFile), []byte(taskGo), 0o644)
}

File: templates/ts-bun-aoc/lib/get-aoc-input/scaffold.ts
import path from 'node:path';

interface WriteTaskInput {
  data: string;
  outputDir: string;
}
export async function writeTask({ data, outputDir }: WriteTaskInput): Promise<void> {
  const taskInputPath = path.join(outputDir, 'input.txt');
  await Bun.write(taskInputPath, data);
}

interface WriteTemplateInput {
  year: number;
  day: number;
  outputDir: string;
  title: string;
}
export async function writeTemplate({ year, day, outputDir, title }: WriteTemplateInput): Promise<void> {
  const README_TEMPLATE = `[${title}](https://adventofcode.com/${year}/day/${day} "${title}")

\`\`\`shell
bun run task.ts
\`\`\`
`;
  const readmePath = path.join(outputDir, 'README.md');
  await Bun.write(readmePath, README_TEMPLATE);

  const TASK_TEMPLATE = `import { parseInput } from '../../lib/parse-input';

function part1(input: string[]) {}

function part2(input: string[]) {}

async function go(): Promise<void> {
  console.time('task');

  console.time('parse-input');
  const input = await parseInput('./input.txt');
  console.timeEnd('parse-input');

  console.time('part 1');
  const res1 = part1(input);
  console.log('PART 1: ', res1);
  console.timeEnd('part 1');

  console.time('part 2');
  const res2 = part2(input);
  console.log('PART 2: ', res2);
  console.timeEnd('part 2');

  console.timeEnd('task');
}

void go();
`;
  const taskPath = path.join(outputDir, 'task.ts');
  await Bun.write(taskPath, TASK_TEMPLATE);
}
